File: write_to_lcd/vbnet/WeatherStation.vb
Imports Tinkerforge

Module WeatherStation
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223

    Private ipcon As IPConnection = Nothing
    Private brickletLCD As BrickletLCD20x4 = Nothing
    Private brickletAmbientLight As BrickletAmbientLight = Nothing
    Private brickletAmbientLightV2 As BrickletAmbientLightV2 = Nothing
    Private brickletAmbientLightV3 As BrickletAmbientLightV3 = Nothing
    Private brickletHumidity As BrickletHumidity = Nothing
    Private brickletHumidityV2 As BrickletHumidityV2 = Nothing
    Private brickletBarometer As BrickletBarometer = Nothing
    Private brickletBarometerV2 As BrickletBarometerV2 = Nothing

    Sub IlluminanceCB(ByVal sender As BrickletAmbientLight, ByVal illuminance As Integer)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Illuminanc {0,6:###.00} lx", illuminance/10.0)
            brickletLCD.WriteLine(0, 0, text)
            System.Console.WriteLine("Write to line 0: " + text)
        End If
    End Sub

    Sub IlluminanceV2CB(ByVal sender As BrickletAmbientLightV2, ByVal illuminance As Long)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Illumina {0,8:###.00} lx", illuminance/100.0)
            brickletLCD.WriteLine(0, 0, text)
            System.Console.WriteLine("Write to line 0: " + text)
        End If
    End Sub

    Sub IlluminanceV3CB(ByVal sender As BrickletAmbientLightV3, ByVal illuminance As Long)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Illumina {0,8:###.00} lx", illuminance/100.0)
            brickletLCD.WriteLine(0, 0, text)
            System.Console.WriteLine("Write to line 0: " + text)
        End If
    End Sub

    Sub HumidityCB(ByVal sender As BrickletHumidity, ByVal humidity As Integer)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Humidity   {0,6:###.00} %", humidity/10.0)
            brickletLCD.WriteLine(1, 0, text)
            System.Console.WriteLine("Write to line 1: " + text)
        End If
    End Sub

    Sub HumidityV2CB(ByVal sender As BrickletHumidityV2, ByVal humidity As Integer)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Humidity   {0,6:###.00} %", humidity/100.0)
            brickletLCD.WriteLine(1, 0, text)
            System.Console.WriteLine("Write to line 1: " + text)
        End If
    End Sub

    Sub AirPressureCB(ByVal sender As BrickletBarometer, ByVal airPressure As Integer)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Air Press {0,7:####.00} mb", airPressure/1000.0)
            brickletLCD.WriteLine(2, 0, text)
            System.Console.WriteLine("Write to line 2: " + text)

            Dim temperature As Integer
            Try
                temperature = sender.GetChipTemperature()
            Catch e As TinkerforgeException
                System.Console.WriteLine("Could not get temperature" + e.Message)
                Return
            End Try

            ' &HDF == ° on LCD 20x4 charset
            text = String.Format("Temperature {0,5:##.00} {1}C", temperature/100.0, Chr(&HDF))
            brickletLCD.WriteLine(3, 0, text)
            System.Console.WriteLine("Write to line 3: " + text.Replace(Chr(&HDF), "°"C))
        End If
    End Sub

    Sub AirPressureV2CB(ByVal sender As BrickletBarometerV2, ByVal airPressure As Integer)
        If brickletLCD IsNot Nothing Then
            Dim text As String = String.Format("Air Press {0,7:####.00} mb", airPressure/1000.0)
            brickletLCD.WriteLine(2, 0, text)
            System.Console.WriteLine("Write to line 2: " + text)

            Dim temperature As Integer
            Try
                temperature = sender.GetTemperature()
            Catch e As TinkerforgeException
                System.Console.WriteLine("Could not get temperature" + e.Message)
                Return
            End Try

            ' &HDF == ° on LCD 20x4 charset
            text = String.Format("Temperature {0,5:##.00} {1}C", temperature/100.0, Chr(&HDF))
            brickletLCD.WriteLine(3, 0, text)
            System.Console.WriteLine("Write to line 3: " + text.Replace(Chr(&HDF), "°"C))
        End If
    End Sub

    Sub EnumerateCB(ByVal sender As IPConnection, ByVal uid As String, _
                    ByVal connectedUid As String, ByVal position As Char, _
                    ByVal hardwareVersion() As Short, ByVal firmwareVersion() As Short, _
                    ByVal deviceIdentifier As Integer, ByVal enumerationType As Short)
        If enumerationType = IPConnection.ENUMERATION_TYPE_CONNECTED Or _
           enumerationType = IPConnection.ENUMERATION_TYPE_AVAILABLE Then
            If deviceIdentifier = BrickletLCD20x4.DEVICE_IDENTIFIER Then
                Try
                    brickletLCD = New BrickletLCD20x4(UID, ipcon)
                    brickletLCD.ClearDisplay()
                    brickletLCD.BacklightOn()
                    System.Console.WriteLine("LCD 20x4 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("LCD 20x4 init failed: " + e.Message)
                    brickletLCD = Nothing
                End Try
            Else If deviceIdentifier = BrickletAmbientLight.DEVICE_IDENTIFIER Then
                Try
                    brickletAmbientLight = New BrickletAmbientLight(UID, ipcon)
                    brickletAmbientLight.SetIlluminanceCallbackPeriod(1000)
                    AddHandler brickletAmbientLight.Illuminance, AddressOf IlluminanceCB
                    System.Console.WriteLine("Ambient Light initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Ambient Light init failed: " + e.Message)
                    brickletAmbientLight = Nothing
                End Try
            Else If deviceIdentifier = BrickletAmbientLightV2.DEVICE_IDENTIFIER Then
                Try
                    brickletAmbientLightV2 = New BrickletAmbientLightV2(UID, ipcon)
                    brickletAmbientLightV2.SetConfiguration(BrickletAmbientLightV2.ILLUMINANCE_RANGE_64000LUX, _
                                                            BrickletAmbientLightV2.INTEGRATION_TIME_200MS)
                    brickletAmbientLightV2.SetIlluminanceCallbackPeriod(1000)
                    AddHandler brickletAmbientLightV2.Illuminance, AddressOf IlluminanceV2CB
                    System.Console.WriteLine("Ambient Light 2.0 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Ambient Light 2.0 init failed: " + e.Message)
                    brickletAmbientLightV2 = Nothing
                End Try
            Else If deviceIdentifier = BrickletAmbientLightV3.DEVICE_IDENTIFIER Then
                Try
                    brickletAmbientLightV3 = New BrickletAmbientLightV3(UID, ipcon)
                    brickletAmbientLightV3.SetConfiguration(BrickletAmbientLightV3.ILLUMINANCE_RANGE_64000LUX, _
                                                            BrickletAmbientLightV3.INTEGRATION_TIME_200MS)
                    brickletAmbientLightV3.SetIlluminanceCallbackConfiguration(1000, True, "x"C, 0, 0)
                    AddHandler brickletAmbientLightV3.IlluminanceCallback, AddressOf IlluminanceV3CB
                    System.Console.WriteLine("Ambient Light 3.0 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Ambient Light 3.0 init failed: " + e.Message)
                    brickletAmbientLightV3 = Nothing
                End Try
            Else If deviceIdentifier = BrickletHumidity.DEVICE_IDENTIFIER Then
                Try
                    brickletHumidity = New BrickletHumidity(UID, ipcon)
                    brickletHumidity.SetHumidityCallbackPeriod(1000)
                    AddHandler brickletHumidity.Humidity, AddressOf HumidityCB
                    System.Console.WriteLine("Humidity initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Humidity init failed: " + e.Message)
                    brickletHumidity = Nothing
                End Try
            Else If deviceIdentifier = BrickletHumidityV2.DEVICE_IDENTIFIER Then
                Try
                    brickletHumidityV2 = New BrickletHumidityV2(UID, ipcon)
                    brickletHumidityV2.SetHumidityCallbackConfiguration(1000, True, "x"C, 0, 0)
                    AddHandler brickletHumidityV2.HumidityCallback, AddressOf HumidityV2CB
                    System.Console.WriteLine("Humidity 2.0 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Humidity 2.0 init failed: " + e.Message)
                    brickletHumidityV2 = Nothing
                End Try
            Else If deviceIdentifier = BrickletBarometer.DEVICE_IDENTIFIER Then
                Try
                    brickletBarometer = New BrickletBarometer(UID, ipcon)
                    brickletBarometer.SetAirPressureCallbackPeriod(1000)
                    AddHandler brickletBarometer.AirPressure, AddressOf AirPressureCB
                    System.Console.WriteLine("Barometer initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Barometer init failed: " + e.Message)
                    brickletBarometer = Nothing
                End Try
            Else If deviceIdentifier = BrickletBarometerV2.DEVICE_IDENTIFIER Then
                Try
                    brickletBarometerV2 = New BrickletBarometerV2(UID, ipcon)
                    brickletBarometerV2.SetAirPressureCallbackConfiguration(1000, True, "x"C, 0, 0)
                    AddHandler brickletBarometerV2.AirPressureCallback, AddressOf AirPressureV2CB
                    System.Console.WriteLine("Barometer 2.0 initialized")
                Catch e As TinkerforgeException
                    System.Console.WriteLine("Barometer 2.0 init failed: " + e.Message)
                    brickletBarometerV2 = Nothing
                End Try
            End If
        End If
    End Sub

    Sub ConnectedCB(ByVal sender As IPConnection, ByVal connectedReason as Short)
        If connectedReason = IPConnection.CONNECT_REASON_AUTO_RECONNECT Then
            System.Console.WriteLine("Auto Reconnect")
            while True
                Try
                    ipcon.Enumerate()
                    Exit While
                Catch e As NotConnectedException
                    System.Console.WriteLine("Enumeration Error: " + e.Message)
                    System.Threading.Thread.Sleep(1000)
                End Try
            End While
        End If
    End Sub

    Sub Main()
        ipcon = New IPConnection()
        while True
            Try
                ipcon.Connect(HOST, PORT)
                Exit While
            Catch e As System.Net.Sockets.SocketException
                System.Console.WriteLine("Connection Error: " + e.Message)
                System.Threading.Thread.Sleep(1000)
            End Try
        End While

        AddHandler ipcon.EnumerateCallback, AddressOf EnumerateCB
        AddHandler ipcon.Connected, AddressOf ConnectedCB

        while True
            try
                ipcon.Enumerate()
                Exit While
            Catch e As NotConnectedException
                System.Console.WriteLine("Enumeration Error: " + e.Message)
                System.Threading.Thread.Sleep(1000)
            End Try
        End While

        System.Console.WriteLine("Press key to exit")
        System.Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: write_to_lcd/delphi/WeatherStation.pas
program WeatherStation;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  Device,
  SysUtils,
  IPConnection,
  BrickletLCD20x4,
  BrickletHumidity,
  BrickletBarometer,
  BrickletHumidityV2,
  BrickletBarometerV2,
  BrickletAmbientLight,
  BrickletAmbientLightV2,
  BrickletAmbientLightV3;

const
  HOST = 'localhost';
  PORT = 4223;

type
  TWeatherStation = class

  private
    ipcon: TIPConnection;
    brickletLCD: TBrickletLCD20x4;
    brickletHumidity: TBrickletHumidity;
    brickletBarometer: TBrickletBarometer;
    brickletHumidityV2: TBrickletHumidityV2;
    brickletBarometerV2: TBrickletBarometerV2;
    brickletAmbientLight: TBrickletAmbientLight;
    brickletAmbientLightV2: TBrickletAmbientLightV2;
    brickletAmbientLightV3: TBrickletAmbientLightV3;

  public
    procedure Execute;
    constructor Create;
    destructor Destroy; override;
    procedure EnumerateCB(sender: TIPConnection;
                          const uid: string;
                          const connectedUid: string;
                          const position: char;
                          const hardwareVersion: TVersionNumber;
                          const firmwareVersion: TVersionNumber;
                          const deviceIdentifier: word;
                          const enumerationType: byte);
    procedure HumidityCB(sender: TBrickletHumidity; const humidity: word);
    procedure HumidityV2CB(sender: TBrickletHumidityV2; const humidity: word);
    procedure ConnectedCB(sender: TIPConnection; const connectedReason: byte);
    procedure AirPressureCB(sender: TBrickletBarometer; const airPressure: longint);
    procedure IlluminanceCB(sender: TBrickletAmbientLight; const illuminance: word);
    procedure AirPressureV2CB(sender: TBrickletBarometerV2; const airPressure: longint);
    procedure IlluminanceV2CB(sender: TBrickletAmbientLightV2; const illuminance: longword);
    procedure IlluminanceV3CB(sender: TBrickletAmbientLightV3; const illuminance: longword);
  end;

var
  ws: TWeatherStation;

constructor TWeatherStation.Create;
begin
  ipcon := nil;
  brickletLCD := nil;
  brickletHumidity := nil;
  brickletBarometer := nil;
  brickletHumidityV2 := nil;
  brickletBarometerV2 := nil;
  brickletAmbientLight := nil;
  brickletAmbientLightV2 := nil;
  brickletAmbientLightV3 := nil;
end;

destructor TWeatherStation.Destroy;
begin
  if (ipcon <> nil) then ipcon.Destroy;
  if (brickletLCD <> nil) then brickletLCD.Destroy;
  if (brickletHumidity <> nil) then brickletHumidity.Destroy;
  if (brickletBarometer <> nil) then brickletBarometer.Destroy;
  if (brickletHumidityV2 <> nil) then brickletHumidityV2.Destroy;
  if (brickletBarometerV2 <> nil) then brickletBarometerV2.Destroy;
  if (brickletAmbientLight <> nil) then brickletAmbientLight.Destroy;
  if (brickletAmbientLightV2 <> nil) then brickletAmbientLightV2.Destroy;
  if (brickletAmbientLightV3 <> nil) then brickletAmbientLightV3.Destroy;

  inherited Destroy;
end;

procedure TWeatherStation.ConnectedCB(sender: TIPConnection; const connectedReason: byte);
begin
  if (connectedReason = IPCON_CONNECT_REASON_AUTO_RECONNECT) then
  begin
    WriteLn('Auto Reconnect');

    while (true) do begin
      try
        ipcon.Enumerate;

        break;
      except
        on e: Exception do
        begin
          WriteLn('Enumeration Error: ' + e.Message);
          Sleep(1000);
        end;
      end;
    end;
  end;
end;

procedure TWeatherStation.EnumerateCB(sender: TIPConnection; const uid: string;
                                      const connectedUid: string; const position: char;
                                      const hardwareVersion: TVersionNumber;
                                      const firmwareVersion: TVersionNumber;
                                      const deviceIdentifier: word; const enumerationType: byte);
begin
  if ((enumerationType = IPCON_ENUMERATION_TYPE_CONNECTED) or
      (enumerationType = IPCON_ENUMERATION_TYPE_AVAILABLE)) then
    begin
      if (deviceIdentifier = BRICKLET_LCD_20X4_DEVICE_IDENTIFIER) then
      begin
        try
          brickletLCD := TBrickletLCD20x4.Create(UID, ipcon);
          brickletLCD.ClearDisplay;
          brickletLCD.BacklightOn;
          WriteLn('LCD 20x4 initialized');
        except
          on e: Exception do
          begin
            WriteLn('LCD 20x4 init failed: ' + e.Message);
            brickletLCD := nil;
          end;
        end;
      end
    else if (deviceIdentifier = BRICKLET_AMBIENT_LIGHT_DEVICE_IDENTIFIER) then
    begin
      try
        brickletAmbientLight := TBrickletAmbientLight.Create(uid, ipcon);
        brickletAmbientLight.SetIlluminanceCallbackPeriod(1000);
        brickletAmbientLight.OnIlluminance := {$ifdef FPC}@{$endif}IlluminanceCB;
        WriteLn('Ambient Light initialized');
      except
        on e: Exception do
        begin
          WriteLn('Ambient Light init failed: ' + e.Message);
          brickletAmbientLight := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_AMBIENT_LIGHT_V2_DEVICE_IDENTIFIER) then
    begin
      try
        brickletAmbientLightV2 := TBrickletAmbientLightV2.Create(uid, ipcon);
        brickletAmbientLightV2.SetConfiguration(BRICKLET_AMBIENT_LIGHT_V2_ILLUMINANCE_RANGE_64000LUX,
                                                BRICKLET_AMBIENT_LIGHT_V2_INTEGRATION_TIME_200MS);
        brickletAmbientLightV2.SetIlluminanceCallbackPeriod(1000);
        brickletAmbientLightV2.OnIlluminance := {$ifdef FPC}@{$endif}IlluminanceV2CB;
        WriteLn('Ambient Light 2.0 initialized');
      except
        on e: Exception do
        begin
          WriteLn('Ambient Light 2.0 init failed: ' + e.Message);
          brickletAmbientLightV2 := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_AMBIENT_LIGHT_V3_DEVICE_IDENTIFIER) then
    begin
      try
        brickletAmbientLightV3 := TBrickletAmbientLightV3.Create(uid, ipcon);
        brickletAmbientLightV3.SetConfiguration(BRICKLET_AMBIENT_LIGHT_V2_ILLUMINANCE_RANGE_64000LUX,
                                                BRICKLET_AMBIENT_LIGHT_V2_INTEGRATION_TIME_200MS);
        brickletAmbientLightV3.SetIlluminanceCallbackConfiguration(1000, False, 'x', 0, 0);
        brickletAmbientLightV3.OnIlluminance := {$ifdef FPC}@{$endif}IlluminanceV3CB;
        WriteLn('Ambient Light 3.0 initialized');
      except
        on e: Exception do
        begin
          WriteLn('Ambient Light 3.0 init failed: ' + e.Message);
          brickletAmbientLightV3 := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_HUMIDITY_DEVICE_IDENTIFIER) then
    begin
      try
        brickletHumidity := TBrickletHumidity.Create(uid, ipcon);
        brickletHumidity.SetHumidityCallbackPeriod(1000);
        brickletHumidity.OnHumidity := {$ifdef FPC}@{$endif}HumidityCB;
        WriteLn('Humidity initialized');
      except
        on e: Exception do
        begin
          WriteLn('Humidity init failed: ' + e.Message);
          brickletHumidity := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_HUMIDITY_V2_DEVICE_IDENTIFIER) then
    begin
      try
        brickletHumidityV2 := TBrickletHumidityV2.Create(uid, ipcon);
        brickletHumidityV2.SetHumidityCallbackConfiguration(1000, true, 'x', 0, 0);
        brickletHumidityV2.OnHumidity := {$ifdef FPC}@{$endif}HumidityV2CB;
        WriteLn('Humidity 2.0 initialized');
      except
        on e: Exception do
        begin
          WriteLn('Humidity 2.0 init failed: ' + e.Message);
          brickletHumidityV2 := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_BAROMETER_DEVICE_IDENTIFIER) then
    begin
      try
        brickletBarometer := TBrickletBarometer.Create(uid, ipcon);
        brickletBarometer.SetAirPressureCallbackPeriod(1000);
        brickletBarometer.OnAirPressure := {$ifdef FPC}@{$endif}AirPressureCB;
        WriteLn('Barometer initialized');
      except
        on e: Exception do
        begin
          WriteLn('Barometer init failed: ' + e.Message);
          brickletBarometer := nil;
        end;
      end;
    end
    else if (deviceIdentifier = BRICKLET_BAROMETER_V2_DEVICE_IDENTIFIER) then
    begin
      try
        brickletBarometerV2 := TBrickletBarometerV2.Create(uid, ipcon);
        brickletBarometerV2.SetAirPressureCallbackConfiguration(1000, False, 'x', 0, 0);
        brickletBarometerV2.OnAirPressure := {$ifdef FPC}@{$endif}AirPressureV2CB;
        WriteLn('Barometer 2.0 initialized');
      except
        on e: Exception do
        begin
          WriteLn('Barometer 2.0 init failed: ' + e.Message);
          brickletBarometerV2 := nil;
        end;
      end;
    end;
  end;
end;

procedure TWeatherStation.IlluminanceCB(sender: TBrickletAmbientLight; const illuminance: word);

var text: string;

begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Illuminanc %6.2f lx', [illuminance/10.0]);
    brickletLCD.WriteLine(0, 0, text);
    WriteLn('Write to line 0: ' + text);
  end;
end;

procedure TWeatherStation.IlluminanceV2CB(sender: TBrickletAmbientLightV2;
                                          const illuminance: longword);

var text: string;

begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Illumina %8.2f lx', [illuminance/100.0]);
    brickletLCD.WriteLine(0, 0, text);
    WriteLn('Write to line 0: ' + text);
  end;
end;

procedure TWeatherStation.IlluminanceV3CB(sender: TBrickletAmbientLightV3;
                                          const illuminance: longword);
var text: string;
begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Illumina %8.2f lx', [illuminance / 100.0]);
    brickletLCD.WriteLine(0, 0, text);
    WriteLn('Write to line 0: ' + text);
  end;
end;

procedure TWeatherStation.HumidityCB(sender: TBrickletHumidity;
                                     const humidity: word);
var text: string;
begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Humidity   %6.2f %%', [humidity/10.0]);
    brickletLCD.WriteLine(1, 0, text);
    WriteLn('Write to line 1: ' + text);
  end;
end;

procedure TWeatherStation.HumidityV2CB(sender: TBrickletHumidityV2;
                                       const humidity: word);
var text: string;
begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Humidity   %6.2f %%', [humidity/100.0]);
    brickletLCD.WriteLine(1, 0, text);
    WriteLn('Write to line 1: ' + text);
  end;
end;

procedure TWeatherStation.AirPressureCB(sender: TBrickletBarometer;
                                        const airPressure: longint);
var text: string; temperature: smallint;
begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Air Press %7.2f mb', [airPressure/1000.0]);
    brickletLCD.WriteLine(2, 0, text);
    WriteLn('Write to line 2: ' + text);

    try
      temperature := brickletBarometer.GetChipTemperature;
    except
      on e: Exception do
      begin
        WriteLn('Could not get temperature: ' + e.Message);
        exit;
      end;
    end;

    { $DF == ° on LCD 20x4 charset. }
    text := Format('Temperature %5.2f %sC', [temperature/100.0, '' + char($DF)]);
    brickletLCD.WriteLine(3, 0, text);
    text := StringReplace(text, char($DF), '°', [rfReplaceAll]);
    WriteLn('Write to line 3: ' + text);
  end;
end;

procedure TWeatherStation.AirPressureV2CB(sender: TBrickletBarometerV2;
                                          const airPressure: longint);
var text: string;
temperature: smallint;
begin
  if (brickletLCD <> nil) then
  begin
    text := Format('Air Press %7.2f mb', [airPressure/1000.0]);
    brickletLCD.WriteLine(2, 0, text);
    WriteLn('Write to line 2: ' + text);

    try
      temperature := brickletBarometerV2.GetTemperature;
    except
      on e: Exception do
      begin
        WriteLn('Could not get temperature: ' + e.Message);
        exit;
      end;
    end;

    { $DF == ° on LCD 20x4 charset. }
    text := Format('Temperature %5.2f %sC', [temperature/100.0, '' + char($DF)]);
    brickletLCD.WriteLine(3, 0, text);
    text := StringReplace(text, char($DF), '°', [rfReplaceAll]);
    WriteLn('Write to line 3: ' + text);
  end;
end;

procedure TWeatherStation.Execute;
begin
  ipcon := TIPConnection.Create;
  while (true) do
  begin
    try
      ipcon.Connect(HOST, PORT);
      break;
    except
      on e: Exception do
      begin
        WriteLn('Connection Error: ' + e.Message);
        Sleep(1000);
      end;
    end;
  end;
  ipcon.OnEnumerate := {$ifdef FPC}@{$endif}EnumerateCB;
  ipcon.OnConnected := {$ifdef FPC}@{$endif}ConnectedCB;
  while (true) do
  begin
    try
      ipcon.Enumerate;
      break;
    except
      on e: Exception do
      begin
        WriteLn('Enumeration Error: ' + e.Message);
        Sleep(1000);
      end;
    end;
  end;
  WriteLn('Press key to exit');
  ReadLn;
end;

begin
  ws := TWeatherStation.Create;
  ws.Execute;
  ws.Destroy;
end.
